### PR TITLE
expose Vehicle.spin_sprite to Car scripting

### PR DESCRIPTION
**Pull request by guysv**
Demo:

https://github.com/user-attachments/assets/19a42399-170d-4a1c-ab96-3d2a789e35f8

### CODE REVIEW

**Comment by duncanspumpkin** [comment]
Is this really the name we want to expose here. The way i see it a sprite is an individual image this is more an index into a set of images that changes depending on the pitch. @Sadret do you have an thoughts on this.

**Comment by Sadret** [comment]
I needed a bit to understand the conceptual difference between yaw and spin (rotations are applied in the order spin - bank - pitch - yaw, which makes spin and yaw different). Compared to bank, pitch and yaw, it is the case that spin is only a cosmetic property and does not have any impact on game mechanisms or physics.
That said, I do not think this fact should be exposed to the API. Therefore I am in favour of just calling it spin. The comment then should explain the value range (0-255).
Connected to this, I want to bring up the discussion in #22048. While pitch and bank will be represented by enums and are working weird in their ways, yaw is planned to be a number together with a multiplier (or the inverse of it): "entity's rotation is yaw*360/numYawValues", where numYawValues is currently 32 but supposed to be raised in some unknown future. I think that here we have a similar situation, in that there is one byte to represent the spin but only 5 bit of them are used (in the video you can see that the sprite changes only after a step of 8). But I also want to say that I am not a proponent of this suggestion.
Lastly, let's bring in @Gymnasiast and @Basssiiie.

**Comment by guysv** [comment]
You can check out the codebase usages there are basically 6 interesting uses for that variable. they appear all to be cosmetic with no real effect on other gameplay mechanics. I could rename it to 'Spin' (or Rotation might be more adequate?) but personalIy believe the API should not obstruct the inner workings of the game.

Let's be honest here, if we are already writing a 'trainer' API for the game, why should we mask information from the plugin authors? It only makes it harder to delve into the game source when in need 😂 .

**Comment by guysv** [comment]
The point about the unused bits that should be masked is probably valid though.

**Comment by Sadret** [comment]
> but personalIy believe the API should not obstruct the inner workings of the game.
> Let's be honest here, if we are already writing a 'trainer' API for the game, why should we mask information from the plugin authors?

We usually have the opposite philosophy here, where we try not to leak implementation details into the API. Whenever it happens that implementations changes (which has happened in the past), we can usually do it without needing to change the API (which is highly desirable). For example, maybe we decide that the spin should affect the ride ratings or force statistics. Then, the spin would become more than just cosmetics and either we would need to change the API or the naming would stop making sense.

**Comment by duncanspumpkin** [comment]
@guysv there are many variables in the codebase that were named before we had fully settled on what we want to call them. The scripting api sort of sets a name in stone as we wouldn't want to break scripts if we decided later to rename. We would be renaming the internal name to match whatever is chosen in this PR.

**Comment by guysv** [comment]
these are indeed valid points yall making. Settled on `rotation` then?

**Comment by guysv** [comment]
These are indeed valid points yall making. Also I now better understand why it's tricky to call it 'rotation' when in the future we'll might wanna add yaw and friends. It's also not 'seat rotation' exactly. that's another property that's used in the multidimension coaster lol. I'm open for suggestions.

**Comment by Sadret** [comment]
I'm fine with `spin`.

File: distribution/changelog.txt [hunk 1]
```diff
@@ -6,7 +6,8 @@
 - Feature: [#22172] [Plugin] Expose ride satisfaction ratings to the plugin API.
 - Feature: [#22184] [Plugin] Expose staff statistics to the plugin API.
 - Feature: [#22213] [Plugin] Allow plugins to focus on textboxes in custom windows.
-- Feature: [#22272] [Plugin] Expose ride vehicle’s current track type via car trackLocation.
+- Feature: [#22272] [Plugin] Expose ride vehicle’s current track type to the plugin API.
```
Comment (Basssiiie) [inline distribution/changelog.txt:9]: This changelog item is not from this PR.

**Review by Basssiiie — CHANGES_REQUESTED** [review]
I also agree with `spin` as the naming for the API property. As per my updated feedback on #22048, it makes better sense with 0-255 as values, since 1 spinning unit != 1 sprite.

Apart this and my other comment below, the implementation looks good to me. 🙂

File: distribution/changelog.txt (hunk 1 above)
Comment (guysv) [reply distribution/changelog.txt:9]: correct. I copied my previous PR number by mistake.

**Comment by guysv** [comment]
I wanna suggest 'spinningProgress'? to align with `spinningInertia` and `spinningFriction` that already exist on `Car`

**Comment by spacek531** [comment]
`spin` is better than `spinningProgress` in my opinion. Where is the spin progressing to? It's just an angle.

**Comment by guysv** [comment]
Sounded obvious to me that the spin progress marks the progress towards one full spin. maybe spinAngle?

**Comment by Basssiiie** [comment]
If I'm not mistaken, while "spin" can be a verb like "spinning", it is also a noun just like pitch/yaw, so just "spin" is fine. The "spin" of a vehicle is how fast it is rotating.

**Comment by guysv** [comment]
alright alight `spin` it is 😆 . will push soon.

**Comment by guysv** [comment]
alright renamed to `spin` and added shifts to spin getter/setter to skip `spin_sprite` reserved bytes. anything else before the merge?

**Comment by duncanspumpkin** [comment]
Just need to increment the scripting api version

**Comment by guysv** [comment]
done.

**Comment by duncanspumpkin** [comment]
I'm just looking at the discussions again. Was everyone in agreement with removing the lower bits? @Basssiiie and @Sadret i think are saying keep it 0-255 but add a comment that at present all vehicles only change spin sprite every 8 spin items.

**Comment by guysv** [comment]
I'm also in favor of just keeping the unused bits exposed to scripting. (both get and set).

**Comment by Sadret** [comment]
I would put this in line with the results of the discussion in https://github.com/OpenRCT2/OpenRCT2/pull/22048 and leave it at 0-255.

**Comment by guysv** [comment]
Reverted back to 0-255. Let's get this in for 0.4.13 shall we?

**Review by duncanspumpkin — APPROVED** [review]

**Review by Basssiiie — APPROVED** [review]
LGTM!

**Comment by guysv** [comment]
do I need to change the changelog now that's this goes into 0.4.14?

**Comment by Sadret** [comment]
yes

**Comment by guysv** [comment]
There it is.

**Pull request merged**